moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,586 @@
+# | user_id       | dee87477683645cb8ca4c2963b950e7e                                                                                                                                       |
+# +---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+
+
+def do_parse_args(config):
+    """Parse args and return a config dict"""
+    parser = argparse.ArgumentParser(description="Generate accounting records for OpenStack instances", epilog="-D and -A are mutually exclusive")
+    parser.add_argument("-v", "--verbose", help="output debugging information", action="store_true")
+    parser.add_argument("-n", "--nostate", help="Skip state messages", action="store_true")
+    parser.add_argument("-c", "--collapse-traits", help="Collapse the traits array to the top level", action="store_true")
+    parser.add_argument("-C", "--config-dir", help="Configuration directory")
+    parser.add_argument("-s", "--start", help="Start time for records", required=False)  # pulls the time of the event from the event list
+    parser.add_argument("-e", "--end", help="End time for records", required=False)  # defaults to current time
+    parser.add_argument("-o", "--outdir", help="Output directory")
+    parser.add_argument("--cloud", help="Using the name/credentials in the cloud config file to connect with OpenStack")
+    parser.add_argument("-d", "--db", help="Database name, only valid for --use-db")
+    parser.add_argument("-f", "--filter-noise", help="Filter out noisy events. Requires panko patch", action="store_true")
+
+    args = parser.parse_args()
+
+    config["loglevel"] = logging.CRITICAL
+
+    if args.collapse_traits:
+        config["collapse_traits"] = True
+
+    if args.filter_noise:
+        config["filter_noise"] = True
+
+    if args.cloud:
+        config["cloud"] = args.cloud
+
+    config["config_dir"] = "."
+    if args.config_dir:
+        config["config_dir"] = args.config_file
+
+    config["start"] = None
+    if args.start:
+        config["start"] = args.start
+
+    config["end"] = datetime.datetime.utcnow().isoformat()
+    if args.end:
+        config["end"] = args.end
+
+    config["outdir"] = "."
+    if args.outdir:
+        config["outdir"] = args.outdir
+
+    if args.verbose:
+        config["loglevel"] = logging.WARNING
+
+    if args.nostate:
+        config["nostate"] = True
+        config["skip_events"].append("compute.instance.exists")
+
+    return config
+
+
+def do_read_config(config):
+    """reads from the config file if present"""
+    try:
+        with open(f"{config['config_dir']}/openstack_reporting.json", "r", encoding="utf-8") as config_fp:
+            newconfig = json.load(config_fp)
+            config.update(newconfig)
+    except IOError:
+        return
+    except ValueError:
+        print("Error, Badly formatted config file")
+        exit()
```

Comment:
Consider replacing these two lines with:

```
exit('Error, badly formatted config file')
```

This ensures that the error is printed on *stderr* rather than *stdout*.